libs/wingc/src/type_check.rs

```diff
@@ -1283,11 +1347,13 @@ impl Types {
 	pub fn maybe_unwrap_inference(&self, t: TypeRef) -> TypeRef {
 		if let Type::Inferred(id) = &*t {
 			if let Some(t) = self.get_inference_by_id(*id) {
-				return self.maybe_unwrap_inference(t);
+				self.maybe_unwrap_inference(t)
+			} else {
```

Comment:
My answer is maybe :)

As currently implemented, it is not possible to avoid inference chains if we include cases of invalid programs:

```js
(arg /*: inference(0)*/) /*: inference(1)*/ => {
  return arg;
  // at this point, we only know that inference(1) -> inference(0)
}
```

I can't think of a valid example at the moment, although I vaguely remember running into a one as well. I'll see what I can do in this PR though, because I would prefer that invariant.

Reply:
Ok so for valid cases, inferences are used but I suppose it's more of an implementation detail of the compiler rather than something that a user situationally triggers. Given that, it should be possible to avoid.

However, also given that invalid cases like the above exist and still need to be modeled, I'm not sure there's much value in trying to figure out a different implementation.